File: leetcode/15-3sum/sol.py
# Input: nums = [-1,0,1,2,-1,-4]
# Output: [[-1,-1,2],[-1,0,1]]

# Explanation:
# nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
# nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
# nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
# The distinct triplets are [-1,0,1] and [-1,-1,2].

# Notice that the order of the output and the order of the triplets does not matter.



class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        """
        Time-Complexity = O(n^2) -> 2Sum is O(n), 2Sum on every itemn => n*n = n^2
        Space-Complexity = ??
        """
        nums.sort() # everytime we sort an array the complexity gets to O(nlogn)
        n = len(nums)
        result = []

        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            target = -nums[i]
            left, right = i + 1, n - 1

            while left < right:
                s = nums[left] + nums[right]

                if s == target:
                    result.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1

                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1

                elif s < target:
                    left += 1
                else:
                    right -= 1

        return result

File: leetcode/209-minimum-size-subarray-sum/sol.cpp
#include <algorithm>
#include <vector>

using namespace std;

// Input: target = 7, nums = [2,3,1,2,4,3]
// Output: 2

class Solution {
public:
  int minSubArrayLen(int target, vector<int> &nums) {
    int nsize();
    int left = 0;
    int sum = 0;
    int minLen = INT_MAX;

    for (int right = 0; right < n; right++) {
      sum += nums[right];

      while (sum >= target) {
        minLen = min(minLen, right - left + 1);
        sum -= nums[left];
        left++;
      }
    }

    return minLen == INT_MAX ? 0 : minLen;
  }
};
